Improve `Ord` violation help

Voultapher opened the pull request:
Recent experience in #128083 showed that the panic message when an Ord violation is detected by the new sort implementations can be confusing. So this PR aims to improve it, together with minor bug fixes in the doc comments for sort*, sort_unstable* and select_nth_unstable*.

Is it possible to get these changes into the 1.81 release? It doesn't change behavior and would greatly help when users encounter this panic for the first time, which they may after upgrading to 1.81.

Tagging @orlp

Voultapher committed:
Improve panic message and surrounding documentation for Ord violations

The new sort implementations have the ability to
detect Ord violations in many cases. This commit
improves the message in a way that should help
users realize what went wrong in their program.

Voultapher committed:
Improve panic sections for sort*, sort_unstable* and select_nth_unstable*

- Move panic information into # Panics section
- Fix mentions of T: Ord that should be compare
- Add missing information

tgross35 commented on library/core/src/slice/sort/shared/smallsort.rs:
```diff
@@ -845,7 +846,21 @@ unsafe fn bidirectional_merge<T: FreezeMarker, F: FnMut(&T, &T) -> bool>(
+    //
+    // The panic message talks about a total order because that's what the Ord documentation talks
+    // about and requires, so as to not confuse users.
+    panic!("user-provided comparison function does not correctly implement a total order");
```
Could you add something like "see the documentation for `Ord`" to hint what to do next?

Voultapher replied:
When thinking of the panic message wording I haven't yet found one that conveys the detail that it could be the Ord impl or comparison function without getting very long. I had hoped that "user-provided comparison function" is enough of a hint to get users to think about where they implemented comparison logic to get them on the right track.

tgross35 commented on library/core/src/slice/mod.rs:
```diff
@@ -3042,15 +3048,14 @@ impl<T> [T] {
     /// Median of Medians using Tukey's Ninther for pivot selection, which guarantees linear runtime
     /// for all inputs.
     ///
-    /// It is typically faster than stable sorting, except in a few special cases, e.g., when the
-    /// slice is nearly fully sorted, where `slice::sort` may be faster.
-    ///
```
Why remove these sections, are they no longer accurate?

Voultapher replied:
These sections make sense for `sort_unstable` for which I originally wrote that part. But for `select_nth_unstable` there is currently no stable counterpart, even then that recommendation is based on properties not relevant for partial sorting. In short this is a copy paste mistake 😇

tgross35 commented:
For all the uses of "total order", could we link somewhere? Either the Wikipedia page, or link somewhere on the `Ord` page. 

Maybe `Ord` should just get a "total order" section that shows an example incorrect implementation since it's not always easy to compare code to equations.

workingjubilee commented on library/alloc/src/slice.rs:
```diff
@@ -304,7 +308,9 @@ impl<T> [T] {
     /// clamps at `self.len() / 2`.
     ///
-    /// If `K: Ord` does not implement a total order, the implementation may panic.
+    /// # Panics
+    ///
+    /// May panic if `K: Ord` does not implement a total order.
```
This seems confusingly worded, to me.
- Ord "by definition" is a total order.
- The name of the type variable is K, but you write `K: Ord` here, suggesting that is somehow different from `K`.

When discussing something bounded by Ord, because the trait defines itself as "implemented for traits that have a total order", you are normally allowed to assume that type has a total order. This is discussing the situation where that assumption is violated. That doesn't really happen for K, because it is a parameter, not a concrete instance. It doesn't happen for Ord, because that's a trait, not its implementation. So I think it also doesn't happen for `K: Ord`, but rather happens for a "fully resolved" instance where K has been substituted with a specific type and a specific implementation of Ord.

So I think you should simply say "the implementation of Ord for K". Discussing it as an `impl Ord for K` seems also fine, since it evokes what matters: the definition site, not the bounding site.

Voultapher replied:
I like "the implementation of Ord for K", done.

Voultapher replied:
done

workingjubilee reviewed (commented):
I already anticipate someone is going to read this note and say "but that can't happen because K is Ord and Ord is a total order". This should be clearer about that case.

workingjubilee commented on library/core/src/slice/sort/shared/smallsort.rs:
```diff
@@ -834,9 +834,10 @@ unsafe fn bidirectional_merge<T: FreezeMarker, F: FnMut(&T, &T) -> bool>(
-        // only fail if the comparison operator fails to be Ord, in which case
-        // we will panic and never access the inconsistent state in dst.
+        // We now should have consumed the full input exactly once. This can only fail if the
+        // user-provided comparison operator fails implements a strict weak ordering as required by
+        // Ord incorrectly, in which case we will panic and never access the inconsistent state in
+        // dst.
```
This involves repetition enough to get a double negation? Maybe triple, depending on how you scan it? It can't possibly literally mean "fails to implement Ord incorrectly"?

Also you later say "comparison function", and we should be consistent (unless this code is never hit by `sort_by`?)

```suggestion
        // user-provided comparison function fails to implement a strict weak ordering.
        // In that case we panic and never access the inconsistent state in dst.
```

workingjubilee commented on library/core/src/slice/sort/shared/smallsort.rs:
```diff
@@ -845,7 +846,21 @@ unsafe fn bidirectional_merge<T: FreezeMarker, F: FnMut(&T, &T) -> bool>(
+    // implementation. They are expected to implement a total order as explained in the Ord
+    // documentation.
+    //
+    // By raising this panic we inform the user, that they have a logic bug in their program. If a
```
We generally don't talk about "raising" a panic.
```suggestion
    // By panicking we inform the user they have a logic bug in their program. If a
```

workingjubilee commented on library/core/src/slice/sort/shared/smallsort.rs:
```diff
@@ -845,7 +846,21 @@ unsafe fn bidirectional_merge<T: FreezeMarker, F: FnMut(&T, &T) -> bool>(
+    // documentation.
+    //
+    // By raising this panic we inform the user, that they have a logic bug in their program. If a
+    // strict weak ordering is not given, the concept of comparison based sorting makes no sense.
+    // E.g.: a < b < c < a
```
"makes no sense" isn't quite right, I think. This could be clearer about why we *choose* to not, instead, e.g. just allow this to happen and give back a somewhat-random permutation. The current algorithm would be memory-unsound if it tried to continue, yes? Would it even terminate? It would be inconsistent at best, right, even if we tried to force it to early-return, such that two `.sort()` calls would permute the list in different ways each time?

Basically, something like "We choose to not pay a zillion costs that would impose a burden on correct Ord impls. Instead, we bail if the comparison function is inconsistent".

workingjubilee replied:
"they probably didn't mean it lol" is also fine, idc.

Voultapher replied:
Thanks for the detailed feedback! I'd say the primary motivation for panicking is to inform the user of a likely unintentional logic bug.

workingjubilee replied:
Yeah, that seemed likely, I just noticed we are making an explicit choice between "do what I said" vs. "do what I mean", here.

It also will be even more justified, I think, by the fact that while `sort` and `sort_by` use the same code path, I _suspect_ we'll find out that `sort_by` hits this panic far more (because it gets more one-off functions that may be poorly tested).

workingjubilee commented on library/core/src/slice/sort/shared/smallsort.rs:
```diff
@@ -845,7 +846,21 @@ unsafe fn bidirectional_merge<T: FreezeMarker, F: FnMut(&T, &T) -> bool>(
+    // strict weak ordering is not given, the concept of comparison based sorting makes no sense.
+    // E.g.: a < b < c < a
+    //
+    // The Ord documentation requires users to implement a total order, arguably that's
+    // unnecessarily strict in the context of sorting. Issues only arise if the weaker requirement
```
Comma splice.
```suggestion
    // The Ord documentation requires users to implement a total order. Arguably that's
    // unnecessarily strict in the context of sorting. Issues only arise if the weaker requirement
```

alternatively
```suggestion
    // The Ord documentation requires users to implement a total order which is
    // unnecessarily strict in the context of sorting. Issues only arise if the weaker requirement
```

Voultapher replied:
done

workingjubilee reviewed (changes requested):
Sorry, noticed some typos also.

Voultapher commented:
> Thanks for the followup, this will be great to have.
> 
> For all the uses of "total order", could we link somewhere? Either the Wikipedia page, or link somewhere on the `Ord` page.
> 
> Maybe `Ord` should just get a "total order" section that shows an example incorrect implementation since it's not always easy to compare code to equations.

Good idea, currently the `Ord` docs links to wikipedia, we could do the same. That said doing so might seem a little intimidating to users unfamiliar with the matter. I like the idea of having concrete code that demonstrates the ideas at play.

Voultapher committed:
Apply review comments

- Use if the implementation of [`Ord`] for `T`
  language
- Link to total order wiki page
- Rework total order help and examples
- Improve language to be more precise and less
  prone to misunderstandings.
- Fix usage of `sort_unstable_by` in `sort_by`
  example
- Fix missing author mention
- Use more consistent example input for sort
- Use more idiomatic assert_eq! in examples
- Use more natural "comparison function" language
  instead of "comparator function"

Voultapher committed:
Hide internal sort module

Voultapher commented:
Changes since last review:

- Use if the implementation of [`Ord`] for `T`
  language
- Link to total order wiki page
- Rework total order help and examples
- Improve language to be more precise and less
  prone to misunderstandings.
- Fix usage of `sort_unstable_by` in `sort_by`
  example
- Fix missing author mention
- Use more consistent example input for sort
- Use more idiomatic assert_eq! in examples
- Use more natural "comparison function" language
  instead of "comparator function"
- Hide sort module

Voultapher commented:
There are something like 12 places that document sort usage, so chances are I missed something, please carefully review for typos and mistakes.

Voultapher commented:
I've also started making changes to the `Ord` docs to better help people understand how to correctly derive and implement them, but that has grown out of scope and I plan on creating a separate PR for it. The sort docs link people to relevant information and surface some of it where sensible, but it's tricky balance between too little information and linking to the relevant places and overloading and duplicating information in the sort docs. I hope the current iteration strikes a decent balance between the two.

workingjubilee commented on library/alloc/src/slice.rs:
```diff
@@ -179,15 +179,25 @@ impl<T> [T] {
-    /// mutability are observed in the input. Same is true if `T: Ord` panics.
+    /// If the implementation of [`Ord`] for `T` does not implement a [total order] the resulting
+    /// order of elements in the slice is unspecified. All original elements will remain in the
+    /// slice and any possible modifications via interior mutability are observed in the input. Same
```
(non-blocking)
Q: Are we promising that we run `Ord::cmp` with each element as an operand at least once, even if there is no total order?

possible A: No, this should be *may* re: mutation, and we are only _really_ promising all elements remain in the slice.

Voultapher replied:
The logic of the comparison function is user-defined, so the behavior could depend on some outside sources of non-determinism i.e. network calls, and with that the effects are unbounded.

I'm not sure I fully understand your question. No we don't promise that we run `cmp` for all elements, that alone for the reasons outlined above wouldn't even be enough to have form of "all the logic". Maybe you are referring to the "any possible modifications" part. I'm open to a better formulation here, what I want to communicate is that if your code was executed you will see the effects of that no matter what. The stdlib has been testing this property for a while, and it's required to for a sound implementation. So I think we should document it. Especially the interplay with panics is non-trivial and even some Rust implementations get it wrong [see](https://github.com/Voultapher/sort-research-rs/blob/main/writeup/sort_safety/text.md#property-analysis).

workingjubilee commented on library/alloc/src/slice.rs:
```diff
@@ -179,15 +179,25 @@ impl<T> [T] {
+    /// If the implementation of [`Ord`] for `T` does not implement a [total order] the resulting
+    /// order of elements in the slice is unspecified. All original elements will remain in the
+    /// slice and any possible modifications via interior mutability are observed in the input. Same
+    /// is true if the implementation of [`Ord`] for `T` panics.
```
(non-blocking)
Q: Is there any minimum amount of elements we promise to mutate before panicking?

Voultapher replied:
Nope.

workingjubilee commented on library/alloc/src/slice.rs:
```diff
@@ -179,15 +179,25 @@ impl<T> [T] {
     /// [`sort_unstable`](slice::sort_unstable). The exception are partially sorted slices, which
     /// may be better served with `slice::sort`.
     ///
+    /// Sorting types that only implement [`PartialOrd`] such as [`f32`] and [`f64`] requires
```
```suggestion
    /// Sorting types that only implement [`PartialOrd`] such as [`f32`] and [`f64`] require
```

workingjubilee commented on library/alloc/src/slice.rs:
```diff
@@ -179,15 +179,25 @@ impl<T> [T] {
     ///
+    /// Sorting types that only implement [`PartialOrd`] such as [`f32`] and [`f64`] requires
+    /// additional precautions. For example Rust defines `NaN != NaN`, which doesn't fulfill the
+    /// reflexivity requirement posed by [`Ord`]. By using an alternative comparison function with
```
(nit) Technically correct phrasing, I would say "imposed by", but I think we can make this phrasing less striking by just saying "of"
```suggestion
    /// reflexivity requirement of [`Ord`]. By using an alternative comparison function with
```

workingjubilee commented on library/alloc/src/slice.rs:
```diff
@@ -179,15 +179,25 @@ impl<T> [T] {
+    /// additional precautions. For example Rust defines `NaN != NaN`, which doesn't fulfill the
+    /// reflexivity requirement posed by [`Ord`]. By using an alternative comparison function with
+    /// [`slice::sort_by`] such as [`f32::total_cmp`] or [`f64::total_cmp`] that defines a [total
+    /// order] users can sort slices containing floating point numbers. Alternatively, if one can
```
Some of them are not numbers. More particularly: this sequence should be hyphenated.
```suggestion
    /// order] users can sort slices containing floating-point values. Alternatively, if one can
```

Jaust saying "floats" is also fine.

workingjubilee commented on library/alloc/src/slice.rs:
```diff
@@ -179,15 +179,25 @@ impl<T> [T] {
+    /// additional precautions. For example Rust defines `NaN != NaN`, which doesn't fulfill the
+    /// reflexivity requirement posed by [`Ord`]. By using an alternative comparison function with
+    /// [`slice::sort_by`] such as [`f32::total_cmp`] or [`f64::total_cmp`] that defines a [total
+    /// order] users can sort slices containing floating point numbers. Alternatively, if one can
+    /// guarantee that all values in the slice are comparable with [`PartialOrd::partial_cmp`] *and*
+    /// the implementation forms a [total order], it's possible to sort the slice with `sort_by(|a,
+    /// b| a.partial_cmp(b).unwrap())`.
```
There are times when using _less_ technical language is more clear. In this case, however, I think we should actually go ahead and say something like
```suggestion
    /// order] users can sort slices containing floating-point values. Alternatively, if all values
    /// in the slice are guaranteed to be in a subset for which [`PartialOrd::partial_cmp`]
    /// forms a [total order], it's possible to sort the slice with 
    /// `sort_by(|a, b| a.partial_cmp(b).unwrap())`.
```

Voultapher replied:
I agree.

workingjubilee commented on library/alloc/src/slice.rs:
```diff
@@ -222,30 +235,19 @@ impl<T> [T] {
-    /// via interior mutability are observed in the input. Same is true if the comparator function
-    /// panics. A total order (for all `a`, `b` and `c`):
+    /// If the comparison function `compare` does not implement a [total order] the resulting order
+    /// of elements in the slice is unspecified. All original elements will remain in the slice and
+    /// any possible modifications via interior mutability are observed in the input. Same is true
+    /// if `compare` panics.
```
(non-blocking)
Same internal mutability Q.

This should maybe should be even clearer about which inputs may exhibit internal mutability given the function passed to `sort_by` is FnMut? But explaining the relationship to the `FnMut` thing isn't necessary given that's a preexisting concern.

Voultapher replied:
I touched on why I think it wouldn't be a bad idea to mention it here https://github.com/rust-lang/rust/pull/128273#discussion_r1702748863. No strong feeling either way. It's a non-trivial property that I'd argue is not obvious even if you know what `FnMut` does. One needs to connect a lot of dots to figure out that the function must maintain that property to be sound. I [didn't](https://youtu.be/3JZAQ4Gsl-g?t=1047) initially. So I could see other users not realizing that this property is a given. It only costs us a sentence, and if the confusion potential is low, I'd say the benefits of mentioning it outweigh the costs.

workingjubilee commented on library/core/src/slice/mod.rs:
```diff
@@ -2884,9 +2885,19 @@ impl<T> [T] {
+    /// is true if the implementation of [`Ord`] for `T` panics.
+    ///
+    /// Sorting types that only implement [`PartialOrd`] such as [`f32`] and [`f64`] requires
+    /// additional precautions. For example Rust defines `NaN != NaN`, which doesn't fulfill the
```
This is an IEEE754 requirement about the normal operator's behavior so there's no need to say "Rust defines".
```suggestion
    /// additional precautions. For example, `NaN != NaN`, which doesn't fulfill the
```

Voultapher replied:
Fair enough, I've changed it to `f32::NAN != f32::NAN` to be actual code, to make the connection to the previously named types more apparent.

workingjubilee reviewed (commented):
This looks good. I noticed some remaining nits. They're just typos or non-blocking concerns because the original wording here was signed-off-on by libs already and they don't necessarily relate directly to the concern of the "total order" requirement.

Voultapher committed:
Apply review comments to PartialOrd section

Voultapher commented:
@workingjubilee thanks for the detailed review so far.

workingjubilee commented:
Okie dokie! I think that everything else is stuff that we can defer to later. Thank you everyone.

@bors r+

bors commented:
:pushpin: Commit 613155c96ab32daf12514e76476132e6b84adaf8 has been approved by `workingjubilee`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=workingjubilee 613155c96ab32daf12514e76476132e6b84adaf8 -->
<!-- homu: {"type":"Approved","sha":"613155c96ab32daf12514e76476132e6b84adaf8","approver":"workingjubilee","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

matthiaskrgr commented:
@bors r-
guess this failed here: https://github.com/rust-lang/rust/pull/128739#issuecomment-2272236312

Voultapher committed:
Fix linkchecker issue

Voultapher commented:
While I can't reproduce the issue with `./x.py --stage 0 test linkchecker html-checker` I made small tweak based on the error message that should hopefully fix the issue.

workingjubilee commented:
@bors r+

bors commented:
:pushpin: Commit 1be60b5d2b6ac569d51abd376e6f04e2fc07692c has been approved by `workingjubilee`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=workingjubilee 1be60b5d2b6ac569d51abd376e6f04e2fc07692c -->
<!-- homu: {"type":"Approved","sha":"1be60b5d2b6ac569d51abd376e6f04e2fc07692c","approver":"workingjubilee","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Voultapher commented:
I see this PR hasn't made its way into the beta yet, what is needed for that?

ChrisDenton commented:
Patience. Merging into beta is a manual process. This PR is labelled as `beta-accepted` so it'll be merged (possibly along with other `beta-accepted` PRs) the next time someone does the backports.
